
# 🌼 기계학습 및 데이터 전처리 - 1차시(24.10.31)

---


## 1. pandas 라이브러리

pandas 라이브러리 설치와 불러오기

- pip install pandas
- import pandas as pd

여기서 import : 불러온다, as : ~로  
-> pd라는 이름으로 pandas 라이브러리를 불러온다.  


In [2]:
# pip install pandas

In [3]:
import pandas as pd

## 2. 데이터프레임
### 2-1. 데이터프레임의 구조
- 표, table과 비슷한 형태
- 리스트, 딕셔너리 등을 이용하여 데이터프레임을 생성할 수 있음
- 구조화된 데이터를 다루는 2차원 테이블 형식의 데이터 구조

In [4]:
data = [
    [1, '수빈', 90],
    [2, '예원', 80],
    [3, '도환', 70]
]

In [5]:
df = pd.DataFrame(data, columns=['ID', '이름', '점수'])

In [6]:
df

ID  이름  점수
0   1  수빈  90
1   2  예원  80
2   3  도환  70

In [7]:
# 딕셔너리로 생성
data = [
    {'ID': 1, '이름': '카리나'},
    {'ID': 2, '이름': '윈터'},
    {'ID': 3, '이름': '닝닝'}
]

In [8]:
df = pd.DataFrame(data)

In [9]:
df

ID   이름
0   1  카리나
1   2   윈터
2   3   닝닝

In [10]:
# 값이 리스트인 딕셔너리로 생성
data = {
    '고객ID': [101, 102, 103, 104, 105],
    '이름': ['해린', '혜인', '하니', '다니엘', '민지'],
    '나이': [18, 16, 19, 19, 20],
    '금액': [25000, 30000, 15000, 50000, 100000],
    '구매횟수': [1, 2, 1, 4, 5]
}

In [11]:
df = pd.DataFrame(data)

In [12]:
df

고객ID   이름  나이      금액  구매횟수
0   101   해린  18   25000     1
1   102   혜인  16   30000     2
2   103   하니  19   15000     1
3   104  다니엘  19   50000     4
4   105   민지  20  100000     5

In [13]:
df = pd.DataFrame(data, index=range(5, 10))
df

고객ID   이름  나이      금액  구매횟수
5   101   해린  18   25000     1
6   102   혜인  16   30000     2
7   103   하니  19   15000     1
8   104  다니엘  19   50000     4
9   105   민지  20  100000     5

In [14]:
df = pd.DataFrame(data, index=['row1', 'row2', 'row3', 'row4', 'row5'])
df

고객ID   이름  나이      금액  구매횟수
row1   101   해린  18   25000     1
row2   102   혜인  16   30000     2
row3   103   하니  19   15000     1
row4   104  다니엘  19   50000     4
row5   105   민지  20  100000     5

### 2-2. 데이터프레임 확인하기
- head : 상위 5개 행 출력
- info : 열과 데이터 타입, 메모리 사용량 등 정보 출력
- describe : 수치형 데이터의 평균, 표준편차, 최솟값, 최댓값 등 기본 통계정보 출력

In [15]:
df.head(3)

고객ID  이름  나이     금액  구매횟수
row1   101  해린  18  25000     1
row2   102  혜인  16  30000     2
row3   103  하니  19  15000     1

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, row1 to row5
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   고객ID    5 non-null      int64 
 1   이름      5 non-null      object
 2   나이      5 non-null      int64 
 3   금액      5 non-null      int64 
 4   구매횟수    5 non-null      int64 
dtypes: int64(4), object(1)
memory usage: 240.0+ bytes


In [17]:
df.describe()

고객ID         나이             금액     구매횟수
count    5.000000   5.000000       5.000000  5.00000
mean   103.000000  18.400000   44000.000000  2.60000
std      1.581139   1.516575   33800.887562  1.81659
min    101.000000  16.000000   15000.000000  1.00000
25%    102.000000  18.000000   25000.000000  1.00000
50%    103.000000  19.000000   30000.000000  2.00000
75%    104.000000  19.000000   50000.000000  4.00000
max    105.000000  20.000000  100000.000000  5.00000

### 2-3. 데이터 선택하기
- loc : 인덱스 라벨을 기준으로 데이터 선택
- iloc : 인덱스 번호를 기준으로 데이터 선택

In [18]:
df['이름']

row1     해린
row2     혜인
row3     하니
row4    다니엘
row5     민지
Name: 이름, dtype: object

In [19]:
df[['이름', '금액']]

이름      금액
row1   해린   25000
row2   혜인   30000
row3   하니   15000
row4  다니엘   50000
row5   민지  100000

In [20]:
# loc, iloc
df.loc['row1']

고객ID      101
이름         해린
나이         18
금액      25000
구매횟수        1
Name: row1, dtype: object

In [21]:
df.iloc[0]

고객ID      101
이름         해린
나이         18
금액      25000
구매횟수        1
Name: row1, dtype: object

In [22]:
df.loc[['row1', 'row3']]

고객ID  이름  나이     금액  구매횟수
row1   101  해린  18  25000     1
row3   103  하니  19  15000     1

In [23]:
df.iloc[[0, 2]]

고객ID  이름  나이     금액  구매횟수
row1   101  해린  18  25000     1
row3   103  하니  19  15000     1

In [24]:
df.loc['row1', ['이름', '금액']]

이름       해린
금액    25000
Name: row1, dtype: object

In [25]:
df.iloc[0, [1, 3]]

이름       해린
금액    25000
Name: row1, dtype: object

In [26]:
# loc는 끝 값을 포함
df.loc['row1':'row3']

고객ID  이름  나이     금액  구매횟수
row1   101  해린  18  25000     1
row2   102  혜인  16  30000     2
row3   103  하니  19  15000     1

In [27]:
# iloc는 끝 값을 포함하지 않기 때문에 +1 필요, 숫자 기반
df.iloc[0:3]

고객ID  이름  나이     금액  구매횟수
row1   101  해린  18  25000     1
row2   102  혜인  16  30000     2
row3   103  하니  19  15000     1

In [28]:
df.loc['row1':'row3', ['이름', '금액', '구매횟수']]

이름     금액  구매횟수
row1  해린  25000     1
row2  혜인  30000     2
row3  하니  15000     1

In [29]:
df.iloc[0:3, [1, 3, 4]]

이름     금액  구매횟수
row1  해린  25000     1
row2  혜인  30000     2
row3  하니  15000     1

### Q. 아래 데이터프레임에서 해당하는 값을 찾아 출력해보세요.

In [30]:
data = {
    '학생' : ['철수', '짱구', '훈이', '유리', '수지', '흰둥이', '짱아', '맹구', '미리', '성아'],
    '국어' : [88, 92, 76, 81, 95, 67, 80, 91, 85, 73],
    '영어' : [78, 85, 92, 88, 76, 84, 90, 81, 89, 80],
    '수학' : [90, 78, 82, 94, 87, 79, 83, 88, 76, 92],
    '사회' : [85, 92, 88, 77, 91, 82, 90, 79, 84, 88],
    '과학' : [91, 79, 88, 85, 82, 90, 87, 94, 83, 86]
}
df = pd.DataFrame(data)
df

학생  국어  영어  수학  사회  과학
0   철수  88  78  90  85  91
1   짱구  92  85  78  92  79
2   훈이  76  92  82  88  88
3   유리  81  88  94  77  85
4   수지  95  76  87  91  82
5  흰둥이  67  84  79  82  90
6   짱아  80  90  83  90  87
7   맹구  91  81  88  79  94
8   미리  85  89  76  84  83
9   성아  73  80  92  88  86

1. 훈이의 수학점수를 출력해보세요

In [31]:
print(f'훈이의 수학 점수는 {df.loc[2, "수학"]}점 입니다.')

훈이의 수학 점수는 82점 입니다.


2. 유리의 영어점수를 출력해보세요

In [32]:
print(f'유리의 영어 점수는 {df.loc[3, "영어"]}점 입니다.')

유리의 영어 점수는 88점 입니다.


3. 학생들의 사회 점수만 보기 위해 해당 열을 선택해 출력해보세요

In [33]:
print(df.loc[:, ['학생', '사회']])

    학생  사회
0   철수  85
1   짱구  92
2   훈이  88
3   유리  77
4   수지  91
5  흰둥이  82
6   짱아  90
7   맹구  79
8   미리  84
9   성아  88


4. 철수부터 수지까지의 영어와 과학 점수만 출력해보세요

In [34]:
print(df.loc[:4, ['학생', '영어', '과학']])

   학생  영어  과학
0  철수  78  91
1  짱구  85  79
2  훈이  92  88
3  유리  88  85
4  수지  76  82


5. 흰둥이부터 성아까지 모든 과목의 점수를 출력해보세요

In [35]:
print(df.loc[5:, :])

    학생  국어  영어  수학  사회  과학
5  흰둥이  67  84  79  82  90
6   짱아  80  90  83  90  87
7   맹구  91  81  88  79  94
8   미리  85  89  76  84  83
9   성아  73  80  92  88  86


### 2-4. 데이터 수정하기
- columns : 전체 컬럼명 변경하기
- rename : 선택하여 컬럼명 변경하기

In [36]:
sample_data = {
    'alphabet': ['A', 'B', 'C'],
    'number': [1, 2, 3],
    'color': ['red', 'orange', 'yellow']
}

sample_df = pd.DataFrame(sample_data)
sample_df

alphabet  number   color
0        A       1     red
1        B       2  orange
2        C       3  yellow

In [37]:
sample_df.columns

Index(['alphabet', 'number', 'color'], dtype='object')

In [38]:
sample_df.columns = ['알파벳', '넘버', '컬러']
sample_df

알파벳  넘버      컬러
0   A   1     red
1   B   2  orange
2   C   3  yellow

In [39]:
sample_df.rename(columns={'넘버': '숫자'})

알파벳  숫자      컬러
0   A   1     red
1   B   2  orange
2   C   3  yellow

In [40]:
sample_df.rename(columns={'넘버': '숫자', '컬러': '색깔'}, inplace=True)  # 변경한 값을 고정
sample_df

알파벳  숫자      색깔
0   A   1     red
1   B   2  orange
2   C   3  yellow

In [41]:
sample_df.loc[1, '알파벳'] = 'D'
sample_df

알파벳  숫자      색깔
0   A   1     red
1   D   2  orange
2   C   3  yellow

In [42]:
sample_df.loc[0, ['알파벳', '색깔']] = ['H', 'purple']
sample_df

알파벳  숫자      색깔
0   H   1  purple
1   D   2  orange
2   C   3  yellow

### 2-5. 데이터 필터링
1. 단일조건으로 데이터 선택
2. 여러 조건을 사용한 데이터 선택
3. 특정 값을 포함하는 데이터 선택
4. 특정 조건을 만족하는 행의 인덱스만 선택

In [43]:
data = {
    '학생' : ['철수', '짱구', '훈이', '유리', '수지', '흰둥이', '짱아', '맹구', '미리', '성아'],
    '국어' : [88, 92, 76, 81, 95, 67, 80, 91, 85, 73],
    '영어' : [78, 85, 92, 88, 76, 84, 90, 81, 89, 80],
    '수학' : [90, 78, 82, 94, 87, 79, 83, 88, 76, 92],
    '사회' : [85, 92, 88, 77, 91, 82, 90, 79, 84, 88],
    '과학' : [91, 79, 88, 85, 82, 90, 87, 94, 83, 86]
}
df = pd.DataFrame(data)
df

학생  국어  영어  수학  사회  과학
0   철수  88  78  90  85  91
1   짱구  92  85  78  92  79
2   훈이  76  92  82  88  88
3   유리  81  88  94  77  85
4   수지  95  76  87  91  82
5  흰둥이  67  84  79  82  90
6   짱아  80  90  83  90  87
7   맹구  91  81  88  79  94
8   미리  85  89  76  84  83
9   성아  73  80  92  88  86

In [44]:
df[df['국어'] >= 80]

학생  국어  영어  수학  사회  과학
0  철수  88  78  90  85  91
1  짱구  92  85  78  92  79
3  유리  81  88  94  77  85
4  수지  95  76  87  91  82
6  짱아  80  90  83  90  87
7  맹구  91  81  88  79  94
8  미리  85  89  76  84  83

In [45]:
# 국어 점수가 80점 이상이고, 사회 점수가 85점인 데이터
# and 조건 '&'
df[(df['국어'] >= 80) & (df['사회'] == 85)]

학생  국어  영어  수학  사회  과학
0  철수  88  78  90  85  91

In [46]:
# 영어 점수가 85점 이상이거나, 수학 점수가 90점 이상인 데이터
# or 조건 '|'
df[(df['영어'] >= 85) | (df['수학'] >= 90)]

학생  국어  영어  수학  사회  과학
0  철수  88  78  90  85  91
1  짱구  92  85  78  92  79
2  훈이  76  92  82  88  88
3  유리  81  88  94  77  85
6  짱아  80  90  83  90  87
8  미리  85  89  76  84  83
9  성아  73  80  92  88  86

In [47]:
# 과학 점수가 91점이거나 86점인 데이터
df[df['과학'].isin([91, 86])]

학생  국어  영어  수학  사회  과학
0  철수  88  78  90  85  91
9  성아  73  80  92  88  86

In [48]:
# 영어 점수가 90점 이상인 데이터의 인덱스만 알고 싶다
df[df['영어'] >= 90].index

Index([2, 6], dtype='int64')

### Q. 아래 데이터프레임에서 문제를 해결해보세요

In [49]:
data2 = {
    '상품명': ['노트북', '모니터', '키보드', '마우스', '헤드셋', 'USB 허브', '프린터', '스피커', '웹캠', '책상'],
    '가격': [1200000, 300000, 50000, 25000, 80000, 15000, 200000, 70000, 40000, 150000],
    '수량': [3, 10, 5, 20, 7, 15, 2, 5, 8, 1],
    '카테고리': ['전자기기', '전자기기', '컴퓨터 주변기기', '컴퓨터 주변기기', '컴퓨터 주변기기', '컴퓨터 주변기기', '사무용품', '컴퓨터 주변기기', '컴퓨터 주변기기', '사무용품']
}
df_shopping = pd.DataFrame(data2)
df_shopping

상품명       가격  수량      카테고리
0     노트북  1200000   3      전자기기
1     모니터   300000  10      전자기기
2     키보드    50000   5  컴퓨터 주변기기
3     마우스    25000  20  컴퓨터 주변기기
4     헤드셋    80000   7  컴퓨터 주변기기
5  USB 허브    15000  15  컴퓨터 주변기기
6     프린터   200000   2      사무용품
7     스피커    70000   5  컴퓨터 주변기기
8      웹캠    40000   8  컴퓨터 주변기기
9      책상   150000   1      사무용품

1. 상품명은 제품명으로, 카테고리는 분류로 컬럼명을 변경해보세요

In [50]:
df_shopping.rename(columns={'상품명': '제품명', '카테고리': '분류'}, inplace=True)
df_shopping

제품명       가격  수량        분류
0     노트북  1200000   3      전자기기
1     모니터   300000  10      전자기기
2     키보드    50000   5  컴퓨터 주변기기
3     마우스    25000  20  컴퓨터 주변기기
4     헤드셋    80000   7  컴퓨터 주변기기
5  USB 허브    15000  15  컴퓨터 주변기기
6     프린터   200000   2      사무용품
7     스피커    70000   5  컴퓨터 주변기기
8      웹캠    40000   8  컴퓨터 주변기기
9      책상   150000   1      사무용품

2. 100000 가격 이상의 상품을 데이터프레임 형태로 출력해보세요

In [51]:
df_shopping[df_shopping['가격'] >= 100000]

제품명       가격  수량    분류
0  노트북  1200000   3  전자기기
1  모니터   300000  10  전자기기
6  프린터   200000   2  사무용품
9   책상   150000   1  사무용품

3. 카테고리가 컴퓨터 주변기기인 데이터의 가격과 수량만 출력해보세요

In [52]:
df_shopping[df_shopping['분류'] == '컴퓨터 주변기기'].loc[:, ['가격', '수량']]

가격  수량
2  50000   5
3  25000  20
4  80000   7
5  15000  15
7  70000   5
8  40000   8

4. 가격이 50000원 이상 300000 이하인 제품의 데이터만 출력해보세요

In [53]:
df_shopping[(df_shopping['가격'] >= 50000) & (df_shopping['가격'] <= 300000)]

제품명      가격  수량        분류
1  모니터  300000  10      전자기기
2  키보드   50000   5  컴퓨터 주변기기
4  헤드셋   80000   7  컴퓨터 주변기기
6  프린터  200000   2      사무용품
7  스피커   70000   5  컴퓨터 주변기기
9   책상  150000   1      사무용품

5. 수량이 10개 이상인 데이터들의 '수량'컬럼명을 '재고수량'으로 수정하여 다른 이름의 데이터프레임으로 저장해서 출력해보세요

In [54]:
df_more = df_shopping[df_shopping['수량'] >= 10]
df_more.rename(columns={'수량': '재고수량'}, inplace=True)
df_more

C:\Users\RMARKET\AppData\Local\Temp\ipykernel_16588\1289566356.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_more.rename(columns={'수량': '재고수량'}, inplace=True)


제품명      가격  재고수량        분류
1     모니터  300000    10      전자기기
3     마우스   25000    20  컴퓨터 주변기기
5  USB 허브   15000    15  컴퓨터 주변기기

In [55]:
df_more2 = df_shopping[df_shopping['수량'] >= 10].rename(columns={'수량': '재고수량'})
df_more2

제품명      가격  재고수량        분류
1     모니터  300000    10      전자기기
3     마우스   25000    20  컴퓨터 주변기기
5  USB 허브   15000    15  컴퓨터 주변기기

## 3. 데이터 전처리
### 3-1. 결측치 처리
- fillna : 결측치 채우기

In [56]:
data_with_nan = {
    '고객ID': [101, 102, 103, 104, 105],
    '이름': ['수연' ,'서윤', '소리', '시은', None],
    '나이': [27, None, 24, 27, 26],
    '구매금액': [25000, 30000, None, 50000, 100000],
    '구매횟수': [1, 2, 1, 4, 5]
}

df_with_nan = pd.DataFrame(data_with_nan)
df_with_nan

고객ID    이름    나이      구매금액  구매횟수
0   101    수연  27.0   25000.0     1
1   102    서윤   NaN   30000.0     2
2   103    소리  24.0       NaN     1
3   104    시은  27.0   50000.0     4
4   105  None  26.0  100000.0     5

In [57]:
df_with_nan.isnull()

고객ID     이름     나이   구매금액   구매횟수
0  False  False  False  False  False
1  False  False   True  False  False
2  False  False  False   True  False
3  False  False  False  False  False
4  False   True  False  False  False

In [58]:
# 각 컬럼 별로 결측치가 몇 개 있는지?
df_with_nan.isnull().sum()

고객ID    0
이름      1
나이      1
구매금액    1
구매횟수    0
dtype: int64

In [59]:
# 총 결측치가 몇 개 있는지?
df_with_nan.isnull().sum().sum()

np.int64(3)

#### 결측치 처리 방법
1. 결측치가 포함된 행 제거
2. 결측치가 포함된 열 제거
3. 결측치를 특정 값으로 대체 (ex. 0)
4. 결측치를 평균값으로 대체

In [60]:
# 결측치가 포함된 행 제거
# inplace=True 혹은 다른 이름의 변수로 저장해줘야 삭제 상태가 저장이 된다
df_drop_row = df_with_nan.dropna()
df_drop_row

고객ID  이름    나이     구매금액  구매횟수
0   101  수연  27.0  25000.0     1
3   104  시은  27.0  50000.0     4

In [61]:
# 결측치가 포함된 열 제거
# axis=0은 행, axis=1은 열
df_drop_column = df_with_nan.dropna(axis=1)
df_drop_column

고객ID  구매횟수
0   101     1
1   102     2
2   103     1
3   104     4
4   105     5

In [62]:
# 결측치를 채우자
# float으로 변환된다
df_fill_zero = df_with_nan.fillna(0)
df_fill_zero

고객ID  이름    나이      구매금액  구매횟수
0   101  수연  27.0   25000.0     1
1   102  서윤   0.0   30000.0     2
2   103  소리  24.0       0.0     1
3   104  시은  27.0   50000.0     4
4   105   0  26.0  100000.0     5

In [63]:
# 결측치를 평균 값으로
df_with_nan['나이'].fillna(df_with_nan['나이'].mean(), inplace=True)
df_with_nan

C:\Users\RMARKET\AppData\Local\Temp\ipykernel_16588\1446259407.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_with_nan['나이'].fillna(df_with_nan['나이'].mean(), inplace=True)


고객ID    이름    나이      구매금액  구매횟수
0   101    수연  27.0   25000.0     1
1   102    서윤  26.0   30000.0     2
2   103    소리  24.0       NaN     1
3   104    시은  27.0   50000.0     4
4   105  None  26.0  100000.0     5

In [64]:
df_with_nan['구매금액'].fillna(df_with_nan['구매금액'].mean(), inplace=True)
df_with_nan

C:\Users\RMARKET\AppData\Local\Temp\ipykernel_16588\570889775.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_with_nan['구매금액'].fillna(df_with_nan['구매금액'].mean(), inplace=True)


고객ID    이름    나이      구매금액  구매횟수
0   101    수연  27.0   25000.0     1
1   102    서윤  26.0   30000.0     2
2   103    소리  24.0   51250.0     1
3   104    시은  27.0   50000.0     4
4   105  None  26.0  100000.0     5

In [65]:
# 문자열 결측치를 채우는 방법? 알수없음, 미상 등으로 할 수 있다

### 3-2. 중복 제거 및 정렬
- drop_duplicates : 중복값 제거
- sort_values : 정렬

In [66]:
data_with_duplicates = {
    '고객ID': [101, 102, 103, 101, 101],
    '이름': ['짱아', '철수', '짱구', '수지', '짱아'],
    '구매금액': [25000, 30000, 25000, 50000, 25000],
    '구매횟수': [1, 2, 2, 4, 1]
}
df_with_duplicates = pd.DataFrame(data_with_duplicates)
df_with_duplicates

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2
3   101  수지  50000     4
4   101  짱아  25000     1

In [67]:
df_with_duplicates.duplicated()

0    False
1    False
2    False
3    False
4     True
dtype: bool

In [68]:
df_with_duplicates.duplicated(subset=['고객ID'])

0    False
1    False
2    False
3     True
4     True
dtype: bool

In [69]:
df_drop_duplicates = df_with_duplicates.drop_duplicates()
df_drop_duplicates

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2
3   101  수지  50000     4

In [70]:
df_drop_dupcol = df_with_duplicates.drop_duplicates(subset=['고객ID'])
df_drop_dupcol

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2

In [71]:
# 정렬
df_drop_duplicates.sort_values(by='구매금액')

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
2   103  짱구  25000     2
1   102  철수  30000     2
3   101  수지  50000     4

In [72]:
df_drop_duplicates.sort_values(by='구매금액', ascending=False, inplace=True)
df_drop_duplicates

C:\Users\RMARKET\AppData\Local\Temp\ipykernel_16588\3600264215.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_drop_duplicates.sort_values(by='구매금액', ascending=False, inplace=True)


고객ID  이름   구매금액  구매횟수
3   101  수지  50000     4
1   102  철수  30000     2
0   101  짱아  25000     1
2   103  짱구  25000     2

In [73]:
# 인덱스를 리셋하여 저장
# drop=True라는 옵션을 주지 않으면 새로운 인덱스가 추가된다
df_no_duplicates = df_drop_duplicates.reset_index(drop=True)
df_no_duplicates

고객ID  이름   구매금액  구매횟수
0   101  수지  50000     4
1   102  철수  30000     2
2   101  짱아  25000     1
3   103  짱구  25000     2

## 4. Groupby를 통한 데이터 필터링

In [74]:
df_no_duplicates[df_no_duplicates['구매횟수'] > 2]

고객ID  이름   구매금액  구매횟수
0   101  수지  50000     4

In [75]:
data = {
    '과일': ['사과', '사과', '바나나', '바나나', '포도', '포도'],
    '판매량': [5, 3, 6, 7, 4, 2]
}

df = pd.DataFrame(data)

In [76]:
df

과일  판매량
0   사과    5
1   사과    3
2  바나나    6
3  바나나    7
4   포도    4
5   포도    2

### groupby
- groupby : 데이터프레임의 데이터를 그룹으로 묶고,각 그룹에 대해 특정 연산을 적용할 수 있도록 하는 기능
- 집계 함수(예: sum(), mean(), count(), max() 등)와 함께 자주 사용


In [77]:
df.groupby('과일')['판매량'].sum()

과일
바나나    13
사과      8
포도      6
Name: 판매량, dtype: int64

In [78]:
df.groupby('과일')['판매량'].mean()

과일
바나나    6.5
사과     4.0
포도     3.0
Name: 판매량, dtype: float64

In [79]:
df.groupby('과일')['판매량'].max()

과일
바나나    7
사과     5
포도     4
Name: 판매량, dtype: int64

In [80]:
df.groupby('과일')['판매량'].count()

과일
바나나    2
사과     2
포도     2
Name: 판매량, dtype: int64

In [81]:
df.groupby('과일')['판매량'].min()

과일
바나나    6
사과     3
포도     2
Name: 판매량, dtype: int64

### Q. 실습문제

In [82]:
data3 = [
    ['김철수', '가전', 120000, 2],
    ['이영희', '가구', 80000, 4],
    ['김철수', '가전', 120000, 2],  # 중복 데이터
    ['박민수', '스포츠', 150000, 1],
    ['최수정', '스포츠', 95000, 3],
    ['한지민', '가전', 130000, 2],
    ['이영희', '가구', 50000, 5],
    ['최수정', '스포츠', 95000, 3]  # 중복 데이터
]

# 데이터프레임 생성
buy_df = pd.DataFrame(data3, columns=['고객명', '상품분류', '금액', '수량'])
buy_df

고객명 상품분류      금액  수량
0  김철수   가전  120000   2
1  이영희   가구   80000   4
2  김철수   가전  120000   2
3  박민수  스포츠  150000   1
4  최수정  스포츠   95000   3
5  한지민   가전  130000   2
6  이영희   가구   50000   5
7  최수정  스포츠   95000   3

In [83]:
# 1. 중복된 행을 제거해보세요
bdf_drop_row = buy_df.drop_duplicates()
bdf_drop_row

고객명 상품분류      금액  수량
0  김철수   가전  120000   2
1  이영희   가구   80000   4
3  박민수  스포츠  150000   1
4  최수정  스포츠   95000   3
5  한지민   가전  130000   2
6  이영희   가구   50000   5

In [84]:
# 2. 금액 컬럼을 중심으로 데이터를 정렬해보세요
sort_bdf = bdf_drop_row.sort_values(by='금액')
sort_bdf = sort_bdf.reset_index(drop=True)
sort_bdf

고객명 상품분류      금액  수량
0  이영희   가구   50000   5
1  이영희   가구   80000   4
2  최수정  스포츠   95000   3
3  김철수   가전  120000   2
4  한지민   가전  130000   2
5  박민수  스포츠  150000   1

In [85]:
# 3. 상품분류 컬럼을 기준으로 수량 합계를 출력해보세요
sort_bdf.groupby('상품분류')['수량'].sum()

상품분류
가구     9
가전     4
스포츠    4
Name: 수량, dtype: int64

In [86]:
# 4. 고객명별 금액의 평균값을 출력해보세요
sort_bdf.groupby('고객명')['금액'].mean()

고객명
김철수    120000.0
박민수    150000.0
이영희     65000.0
최수정     95000.0
한지민    130000.0
Name: 금액, dtype: float64

## 5. 데이터프레임 병합 및 연결
### 5-1. 데이터프레임 병합
- merge : 데이터프레임 병합

In [87]:
data1 = {
    '고객ID': [101, 102, 103, 104, 105],
    '이름': ['짱아', '철수', '짱구', '수지', '유리'],
    '구매금액': [25000, 30000, 25000, 50000, 25000],
    '구매횟수': [1, 2, 2, 4, 1]
}
df1 = pd.DataFrame(data1)
df1

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2
3   104  수지  50000     4
4   105  유리  25000     1

In [88]:
data2 = {
    '고객ID': [101, 102, 103, 104, 105],
    '만족도': [5, 4, 2, 5, 3]
}
df2 = pd.DataFrame(data2)
df2

고객ID  만족도
0   101    5
1   102    4
2   103    2
3   104    5
4   105    3

In [89]:
merged_df =  pd.merge(df1, df2, on='고객ID')
merged_df

고객ID  이름   구매금액  구매횟수  만족도
0   101  짱아  25000     1    5
1   102  철수  30000     2    4
2   103  짱구  25000     2    2
3   104  수지  50000     4    5
4   105  유리  25000     1    3

In [90]:
merged_df['고객ID'] = merged_df['고객ID'].astype('float').astype('int')

In [91]:
merged_df

고객ID  이름   구매금액  구매횟수  만족도
0   101  짱아  25000     1    5
1   102  철수  30000     2    4
2   103  짱구  25000     2    2
3   104  수지  50000     4    5
4   105  유리  25000     1    3

### 5-2. 데이터프레임 연결
- concat : 데이터프레임 세로로 연결

In [92]:
data1 = {
    '고객ID': [101, 102, 103, 104, 105],
    '이름': ['짱아', '철수', '짱구', '수지', '유리'],
    '구매금액': [25000, 30000, 25000, 50000, 25000],
    '구매횟수': [1, 2, 2, 4, 1]
}
df1 = pd.DataFrame(data1)
df1

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2
3   104  수지  50000     4
4   105  유리  25000     1

In [93]:
data2 = {
    '고객ID': [106, 107, 108],
    '이름': ['나라', '민아', '지수'],
    '구매금액': [15000, 10000, 45000],
    '구매횟수': [1, 3, 1]
}
df2 = pd.DataFrame(data2)
df2

고객ID  이름   구매금액  구매횟수
0   106  나라  15000     1
1   107  민아  10000     3
2   108  지수  45000     1

In [94]:
concat_df = pd.concat([df1, df2], ignore_index=True)
concat_df

고객ID  이름   구매금액  구매횟수
0   101  짱아  25000     1
1   102  철수  30000     2
2   103  짱구  25000     2
3   104  수지  50000     4
4   105  유리  25000     1
5   106  나라  15000     1
6   107  민아  10000     3
7   108  지수  45000     1

## 6. 데이터프레임 불러오기, 내보내기
- 주로 excel, csv파일을 불러오는 경우가 많음
- excel : openpyxl 라이브러리 설치 후 read_excel
- csv : 쉼표로 구분한 텍스트 데이터, read_csv

In [95]:
# pip install openpyxl

In [96]:
movie = pd.read_excel('./data/영화데이터.xlsx')
movie

제목      장르      국가    시간    평점
0                베테랑2      액션    대한민국  118분  6.61
1              비긴 어게인     드라마      미국  104분  9.12
2            대도시의 사랑법     드라마    대한민국  118분   NaN
3                  룩백     드라마      일본   57분  9.13
4         에이리언 : 로물루스      공포  영국, 미국  119분  8.74
5             안녕, 할부지   다큐멘터리    대한민국   95분  8.61
6             소년시절의 너  멜로/로맨스      중국  135분  9.05
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108분  9.77
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77분  9.60
9             사랑의 하츄핑   애니메이션    대한민국   86분  8.73
10                파일럿     코미디    대한민국  111분  7.70
11            우리가 끝이야  멜로/로맨스      미국  130분  7.94
12             새벽의 모든     드라마      일본  119분  8.16
13       정국 : 아이 엠 스틸    공연실황    대한민국   94분  9.97
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90분  8.28

In [97]:
concat_df.to_excel('./data/고객정보(인덱스 포함).xlsx')

In [98]:
concat_df.to_excel('./data/고객정보(인덱스 미포함).xlsx', index=False)

In [99]:
foodstruct = pd.read_csv('./data/foodstruct_nutritional_facts.csv')
foodstruct.head()

Food Name Category Name  Calcium  Calories  Omega-3 - DHA  Protein  \
0       Apple        Fruits        6        52            0.0       26   
1     Avocado        Fruits       12       160            0.0        2   
2      Banana        Fruits        5        89            0.0       11   
3  Blackberry        Fruits       29        43            0.0       14   
4   Blueberry        Fruits        6        57            0.0       74   

   Vitamin A  Vitamin B1  Vitamin C  Sugar  Vitamin D  Vitamin E  
0       54.0         2.0         46   10.0        0.0       18.0  
1      146.0         7.0         10   66.0        0.0       21.0  
2       64.0         3.0         87   12.0        0.0        1.0  
3      214.0         2.0         21   49.0        0.0       12.0  
4       54.0         4.0         97   10.0        0.0       57.0

In [100]:
foodstruct.to_csv('./data/foodstruct_out.csv')


## 🌼마무리 문제

아래 문제들을 풀어보세요!

Q1. pandas 라이브러리를 사용하여 이름, 나이, 사는 곳, 취미 등을 포함한 데이터프레임을 생성하고 출력해보세요

In [101]:
info = {
    'name': ['나라', '민아', '윤서'],
    'age': [18, 20, 25],
    'city': ['서울', '경기도', '강원도']
}

infodf = pd.DataFrame(info)
infodf

name  age city
0   나라   18   서울
1   민아   20  경기도
2   윤서   25  강원도

Q2. 영문으로 되어있는 데이터프레임의 컬럼을 한글로 바꾸고, 숫자형으로 되어있는 컬럼을 기준으로 오름차순 정렬하여 출력하세요

In [102]:
infodf.columns = ['이름', '나이', '사는 곳']
infodf.sort_values(by='나이')

이름  나이 사는 곳
0  나라  18   서울
1  민아  20  경기도
2  윤서  25  강원도

Q3. 영화 제목, 장르, 평점 등의 데이터를 담은 엑셀 파일을 불러와서 데이터프레임으로 변환하고, 장르 별 평균 평점을 출력해보세요
* 영화데이터.xlsx

In [103]:
movie = pd.read_excel('./data/영화데이터.xlsx')

movie.groupby('장르')['평점'].mean()

장르
공연실황      9.870000
공포        8.740000
다큐멘터리     8.610000
드라마       8.803333
멜로/로맨스    8.495000
애니메이션     8.870000
액션        6.610000
코미디       7.700000
Name: 평점, dtype: float64

### 마무리 실습

1. 기초 데이터 확인

데이터의 기본 정보를 출력하고, 각 열의 데이터 유형을 확인합니다.  
df.info()와 df.describe()를 활용하여 데이터 요약 정보를 확인하세요.

In [138]:
movie = pd.read_excel('./data/영화데이터.xlsx')
movie

제목      장르      국가    시간    평점
0                베테랑2      액션    대한민국  118분  6.61
1              비긴 어게인     드라마      미국  104분  9.12
2            대도시의 사랑법     드라마    대한민국  118분   NaN
3                  룩백     드라마      일본   57분  9.13
4         에이리언 : 로물루스      공포  영국, 미국  119분  8.74
5             안녕, 할부지   다큐멘터리    대한민국   95분  8.61
6             소년시절의 너  멜로/로맨스      중국  135분  9.05
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108분  9.77
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77분  9.60
9             사랑의 하츄핑   애니메이션    대한민국   86분  8.73
10                파일럿     코미디    대한민국  111분  7.70
11            우리가 끝이야  멜로/로맨스      미국  130분  7.94
12             새벽의 모든     드라마      일본  119분  8.16
13       정국 : 아이 엠 스틸    공연실황    대한민국   94분  9.97
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90분  8.28

In [130]:
# 기본 정보 출력
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   제목      15 non-null     object 
 1   장르      15 non-null     object 
 2   국가      15 non-null     object 
 3   시간      15 non-null     object 
 4   평점      14 non-null     float64
dtypes: float64(1), object(4)
memory usage: 728.0+ bytes


In [131]:
# 데이터 요약 정보
movie.describe()

평점
count  14.000000
mean    8.672143
std     0.897219
min     6.610000
25%     8.190000
50%     8.735000
75%     9.127500
max     9.970000

2. 결측치 처리

각 열에 있는 결측치를 확인하고, 결측치가 있는 열을 출력하세요.  
fillna()를 사용해 결측치를 평균 평점으로 대체하거나, dropna()를 사용해 결측치를 제거하세요.

In [132]:
# 결측치 확인
movie.isnull().sum()

제목    0
장르    0
국가    0
시간    0
평점    1
dtype: int64

In [139]:
# 평균 평점으로 대체
movie_fill = movie.fillna(movie['평점'].mean())

In [140]:
movie_fill

제목      장르      국가    시간        평점
0                베테랑2      액션    대한민국  118분  6.610000
1              비긴 어게인     드라마      미국  104분  9.120000
2            대도시의 사랑법     드라마    대한민국  118분  8.672143
3                  룩백     드라마      일본   57분  9.130000
4         에이리언 : 로물루스      공포  영국, 미국  119분  8.740000
5             안녕, 할부지   다큐멘터리    대한민국   95분  8.610000
6             소년시절의 너  멜로/로맨스      중국  135분  9.050000
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108분  9.770000
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77분  9.600000
9             사랑의 하츄핑   애니메이션    대한민국   86분  8.730000
10                파일럿     코미디    대한민국  111분  7.700000
11            우리가 끝이야  멜로/로맨스      미국  130분  7.940000
12             새벽의 모든     드라마      일본  119분  8.160000
13       정국 : 아이 엠 스틸    공연실황    대한민국   94분  9.970000
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90분  8.280000

In [141]:
# 삭제
movie_drop = movie.dropna(inplace=False)
movie_drop

제목      장르      국가    시간    평점
0                베테랑2      액션    대한민국  118분  6.61
1              비긴 어게인     드라마      미국  104분  9.12
3                  룩백     드라마      일본   57분  9.13
4         에이리언 : 로물루스      공포  영국, 미국  119분  8.74
5             안녕, 할부지   다큐멘터리    대한민국   95분  8.61
6             소년시절의 너  멜로/로맨스      중국  135분  9.05
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108분  9.77
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77분  9.60
9             사랑의 하츄핑   애니메이션    대한민국   86분  8.73
10                파일럿     코미디    대한민국  111분  7.70
11            우리가 끝이야  멜로/로맨스      미국  130분  7.94
12             새벽의 모든     드라마      일본  119분  8.16
13       정국 : 아이 엠 스틸    공연실황    대한민국   94분  9.97
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90분  8.28

3. 데이터 타입 변환

시간 열의 문자열 값(예: '118분')을 정수형으로 변환해 영화 시간(분)을 나타내도록 변경하세요.

In [136]:
movie_fill['시간'] = movie_fill['시간'].str.replace('분', '').astype('int')
movie_fill

제목      장르      국가   시간        평점
0                베테랑2      액션    대한민국  118  6.610000
1              비긴 어게인     드라마      미국  104  9.120000
2            대도시의 사랑법     드라마    대한민국  118  8.672143
3                  룩백     드라마      일본   57  9.130000
4         에이리언 : 로물루스      공포  영국, 미국  119  8.740000
5             안녕, 할부지   다큐멘터리    대한민국   95  8.610000
6             소년시절의 너  멜로/로맨스      중국  135  9.050000
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108  9.770000
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77  9.600000
9             사랑의 하츄핑   애니메이션    대한민국   86  8.730000
10                파일럿     코미디    대한민국  111  7.700000
11            우리가 끝이야  멜로/로맨스      미국  130  7.940000
12             새벽의 모든     드라마      일본  119  8.160000
13       정국 : 아이 엠 스틸    공연실황    대한민국   94  9.970000
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90  8.280000

In [146]:
new_time = []
for i in movie_fill['시간']:
    i = i.replace('분', '')
    new_time.append(int(i))

movie_fill['시간'] = new_time
movie_fill

제목      장르      국가   시간        평점
0                베테랑2      액션    대한민국  118  6.610000
1              비긴 어게인     드라마      미국  104  9.120000
2            대도시의 사랑법     드라마    대한민국  118  8.672143
3                  룩백     드라마      일본   57  9.130000
4         에이리언 : 로물루스      공포  영국, 미국  119  8.740000
5             안녕, 할부지   다큐멘터리    대한민국   95  8.610000
6             소년시절의 너  멜로/로맨스      중국  135  9.050000
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108  9.770000
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77  9.600000
9             사랑의 하츄핑   애니메이션    대한민국   86  8.730000
10                파일럿     코미디    대한민국  111  7.700000
11            우리가 끝이야  멜로/로맨스      미국  130  7.940000
12             새벽의 모든     드라마      일본  119  8.160000
13       정국 : 아이 엠 스틸    공연실황    대한민국   94  9.970000
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90  8.280000

4. 필터링 및 인덱싱

평점이 8.0 이상인 영화들만 선택하고, 해당 데이터만 추출하세요.  
loc나 iloc을 활용하여 특정 행과 열의 데이터를 선택하는 연습을 하세요.

In [112]:
movie_fill[movie_fill['평점'] >= 8.0]

제목      장르      국가   시간        평점
1              비긴 어게인     드라마      미국  104  9.120000
2            대도시의 사랑법     드라마    대한민국  118  8.672143
3                  룩백     드라마      일본   57  9.130000
4         에이리언 : 로물루스      공포  영국, 미국  119  8.740000
5             안녕, 할부지   다큐멘터리    대한민국   95  8.610000
6             소년시절의 너  멜로/로맨스      중국  135  9.050000
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국  108  9.770000
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국   77  9.600000
9             사랑의 하츄핑   애니메이션    대한민국   86  8.730000
12             새벽의 모든     드라마      일본  119  8.160000
13       정국 : 아이 엠 스틸    공연실황    대한민국   94  9.970000
14   극장총집편 봇치 더 록! 전편   애니메이션      일본   90  8.280000

In [113]:
movie_fill.loc[3:5, ['제목', '평점']]

제목    평점
3           룩백  9.13
4  에이리언 : 로물루스  8.74
5      안녕, 할부지  8.61

5. 조건에 따른 필터링

국가가 '대한민국'인 영화 중 평점이 7.0 이상인 영화 목록을 추출하세요.  
여러 조건을 조합하여 데이터를 필터링하고, 결과를 출력하세요.

In [114]:
movie_fill[(movie_fill['국가'] == '대한민국') & (movie_fill['평점'] >= 7.0)]

제목     장르    국가   시간        평점
2            대도시의 사랑법    드라마  대한민국  118  8.672143
5             안녕, 할부지  다큐멘터리  대한민국   95  8.610000
7   임영웅 | 아임 히어로 더 스타   공연실황  대한민국  108  9.770000
8    브레드이발소 : 빵스타의 탄생  애니메이션  대한민국   77  9.600000
9             사랑의 하츄핑  애니메이션  대한민국   86  8.730000
10                파일럿    코미디  대한민국  111  7.700000
13       정국 : 아이 엠 스틸   공연실황  대한민국   94  9.970000

6. 컬럼명 수정 및 정렬

장르 열을 ‘영화장르’로, 시간 열을 ‘상영시간’으로 수정하세요.  
평점을 기준으로 내림차순 정렬하고, 상위 10개의 영화를 출력하세요.

In [115]:
movie_rename = movie_fill.rename(columns={'장르': '영화장르', '시간': '상영시간'})
movie_rename

제목    영화장르      국가  상영시간        평점
0                베테랑2      액션    대한민국   118  6.610000
1              비긴 어게인     드라마      미국   104  9.120000
2            대도시의 사랑법     드라마    대한민국   118  8.672143
3                  룩백     드라마      일본    57  9.130000
4         에이리언 : 로물루스      공포  영국, 미국   119  8.740000
5             안녕, 할부지   다큐멘터리    대한민국    95  8.610000
6             소년시절의 너  멜로/로맨스      중국   135  9.050000
7   임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.770000
8    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.600000
9             사랑의 하츄핑   애니메이션    대한민국    86  8.730000
10                파일럿     코미디    대한민국   111  7.700000
11            우리가 끝이야  멜로/로맨스      미국   130  7.940000
12             새벽의 모든     드라마      일본   119  8.160000
13       정국 : 아이 엠 스틸    공연실황    대한민국    94  9.970000
14   극장총집편 봇치 더 록! 전편   애니메이션      일본    90  8.280000

In [116]:
movie_sort = movie_rename.sort_values(by='평점', ascending=False)
movie_sort = movie_sort.reset_index(drop=True)
movie_sort.head(10)

제목    영화장르      국가  상영시간        평점
0       정국 : 아이 엠 스틸    공연실황    대한민국    94  9.970000
1  임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.770000
2   브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.600000
3                 룩백     드라마      일본    57  9.130000
4             비긴 어게인     드라마      미국   104  9.120000
5            소년시절의 너  멜로/로맨스      중국   135  9.050000
6        에이리언 : 로물루스      공포  영국, 미국   119  8.740000
7            사랑의 하츄핑   애니메이션    대한민국    86  8.730000
8           대도시의 사랑법     드라마    대한민국   118  8.672143
9            안녕, 할부지   다큐멘터리    대한민국    95  8.610000

7. 중복값 제거

영화 제목에 중복된 값이 있는지 확인하고, 중복된 영화가 있다면 제거하세요.

In [117]:
movie_sort.duplicated(subset=['제목'])

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
dtype: bool

In [118]:
movie_dropd = movie_sort.drop_duplicates(subset=['제목'])
movie_dropd

제목    영화장르      국가  상영시간        평점
0        정국 : 아이 엠 스틸    공연실황    대한민국    94  9.970000
1   임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.770000
2    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.600000
3                  룩백     드라마      일본    57  9.130000
4              비긴 어게인     드라마      미국   104  9.120000
5             소년시절의 너  멜로/로맨스      중국   135  9.050000
6         에이리언 : 로물루스      공포  영국, 미국   119  8.740000
7             사랑의 하츄핑   애니메이션    대한민국    86  8.730000
8            대도시의 사랑법     드라마    대한민국   118  8.672143
9             안녕, 할부지   다큐멘터리    대한민국    95  8.610000
10   극장총집편 봇치 더 록! 전편   애니메이션      일본    90  8.280000
11             새벽의 모든     드라마      일본   119  8.160000
12            우리가 끝이야  멜로/로맨스      미국   130  7.940000
13                파일럿     코미디    대한민국   111  7.700000
14               베테랑2      액션    대한민국   118  6.610000

8. 그룹별 통계

장르별로 평점의 평균을 계산하고, 각 장르별 평균 평점을 출력하세요.  
groupby를 사용하여 특정 열에 대한 통계 값을 구하세요.

In [119]:
movie_dropd.groupby('영화장르')['평점'].mean()

영화장르
공연실황      9.870000
공포        8.740000
다큐멘터리     8.610000
드라마       8.770536
멜로/로맨스    8.495000
애니메이션     8.870000
액션        6.610000
코미디       7.700000
Name: 평점, dtype: float64

In [120]:
movie_dropd.groupby('국가')['상영시간'].mean()

국가
대한민국      100.875000
미국        117.000000
영국, 미국    119.000000
일본         88.666667
중국        135.000000
Name: 상영시간, dtype: float64

9. 데이터 병합과 결합

새로운 영화 데이터를 추가하여 concat()을 사용해 기존 데이터프레임에 병합하세요.  
merge()를 활용해 각 영화에 감독 이름과 같은 새로운 정보를 추가해 보세요 (필요시 가상 데이터를 생성해 시도해 볼 수 있습니다).

In [121]:
new_data = pd.DataFrame({
    '제목': ['영화A', '영화B'],
    '영화장르': ['액션', '드라마'],
    '국가': ['대한민국', '미국'],
    '상영시간': [110, 95],
    '평점': [8.5, 7.9]
})

In [122]:
movie_con = pd.concat([movie_dropd, new_data], ignore_index=True)
movie_con

제목    영화장르      국가  상영시간        평점
0        정국 : 아이 엠 스틸    공연실황    대한민국    94  9.970000
1   임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.770000
2    브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.600000
3                  룩백     드라마      일본    57  9.130000
4              비긴 어게인     드라마      미국   104  9.120000
5             소년시절의 너  멜로/로맨스      중국   135  9.050000
6         에이리언 : 로물루스      공포  영국, 미국   119  8.740000
7             사랑의 하츄핑   애니메이션    대한민국    86  8.730000
8            대도시의 사랑법     드라마    대한민국   118  8.672143
9             안녕, 할부지   다큐멘터리    대한민국    95  8.610000
10   극장총집편 봇치 더 록! 전편   애니메이션      일본    90  8.280000
11             새벽의 모든     드라마      일본   119  8.160000
12            우리가 끝이야  멜로/로맨스      미국   130  7.940000
13                파일럿     코미디    대한민국   111  7.700000
14               베테랑2      액션    대한민국   118  6.610000
15                영화A      액션    대한민국   110  8.500000
16                영화B     드라마      미국    95  7.900000

In [123]:
director = {
    '제목': ['정국 : 아이 엠 스틸', '임영웅 | 아임 히어로 더 스타', '브레드이발소 : 빵스타의 탄생', '룩백', '비긴 어게인', '소년시절의 너', '에이리언 : 로물루스', '사랑의 하츄핑'],
    '영화감독': ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']
}

director_df = pd.DataFrame(director)
director_df

제목 영화감독
0       정국 : 아이 엠 스틸    A
1  임영웅 | 아임 히어로 더 스타    B
2   브레드이발소 : 빵스타의 탄생    C
3                 룩백    D
4             비긴 어게인    E
5            소년시절의 너    F
6        에이리언 : 로물루스    G
7            사랑의 하츄핑    H

In [124]:
movie_merge = pd.merge(movie_con, director_df, on='제목')
movie_merge

제목    영화장르      국가  상영시간    평점 영화감독
0       정국 : 아이 엠 스틸    공연실황    대한민국    94  9.97    A
1  임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.77    B
2   브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.60    C
3                 룩백     드라마      일본    57  9.13    D
4             비긴 어게인     드라마      미국   104  9.12    E
5            소년시절의 너  멜로/로맨스      중국   135  9.05    F
6        에이리언 : 로물루스      공포  영국, 미국   119  8.74    G
7            사랑의 하츄핑   애니메이션    대한민국    86  8.73    H

10. 데이터 저장

최종적으로 정리된 데이터를 Excel 또는 CSV 파일로 저장하여, 데이터 변환 및 처리가 완료된 파일을 출력해 보세요.

In [125]:
movie_merge.to_excel('./data/영화데이터_실습.xlsx', index=False)

In [126]:
pd.read_excel('./data/영화데이터_실습.xlsx')

제목    영화장르      국가  상영시간    평점 영화감독
0       정국 : 아이 엠 스틸    공연실황    대한민국    94  9.97    A
1  임영웅 | 아임 히어로 더 스타    공연실황    대한민국   108  9.77    B
2   브레드이발소 : 빵스타의 탄생   애니메이션    대한민국    77  9.60    C
3                 룩백     드라마      일본    57  9.13    D
4             비긴 어게인     드라마      미국   104  9.12    E
5            소년시절의 너  멜로/로맨스      중국   135  9.05    F
6        에이리언 : 로물루스      공포  영국, 미국   119  8.74    G
7            사랑의 하츄핑   애니메이션    대한민국    86  8.73    H